pkg/agent/controller/networkpolicy/networkpolicy_controller_test.go

```diff
@@ -472,6 +539,95 @@ func TestAddNetworkPolicyWithMultipleRules(t *testing.T) {
+	appliedToGroupWatcher.Add(newAppliedToGroup("appliedToGroup1", []v1beta2.GroupMember{*newAppliedToGroupMember("pod1", "ns1")}))
+	appliedToGroupWatcher.Action(watch.Bookmark, nil)
+	for i := 0; i < 2; i++ {
+		select {
+		// The following reconciler.updated events can only come from async rule reconciles,
+		// as the BatchReconcile function of the mock controller has been fault injected.
+		case ruleID := <-reconciler.updated:
+			actualRule, _ := reconciler.getLastRealized(ruleID)
+			if actualRule.Direction == v1beta2.DirectionIn {
+				if !assert.ElementsMatch(t, actualRule.Services, desiredRule1.Services) {
+					t.Errorf("Expected Services %v, got %v", desiredRule1.Services, actualRule.Services)
+				}
+				if !actualRule.FromAddresses.Equal(desiredRule1.FromAddresses) {
+					t.Errorf("Expected FromAddresses %v, got %v", desiredRule1.FromAddresses, actualRule.FromAddresses)
+				}
+				if !actualRule.ToAddresses.Equal(desiredRule1.ToAddresses) {
+					t.Errorf("Expected ToAddresses %v, got %v", desiredRule1.ToAddresses, actualRule.ToAddresses)
+				}
+				if !actualRule.TargetMembers.Equal(desiredRule1.TargetMembers) {
+					t.Errorf("Expected Pods %v, got %v", desiredRule1.TargetMembers, actualRule.TargetMembers)
+				}
+			}
+			if actualRule.Direction == v1beta2.DirectionOut {
+				if !assert.ElementsMatch(t, actualRule.Services, desiredRule2.Services) {
+					t.Errorf("Expected Services %v, got %v", desiredRule2.Services, actualRule.Services)
+				}
+				if !actualRule.FromAddresses.Equal(desiredRule2.FromAddresses) {
+					t.Errorf("Expected FromAddresses %v, got %v", desiredRule2.FromAddresses, actualRule.FromAddresses)
+				}
+				if !actualRule.ToAddresses.Equal(desiredRule2.ToAddresses) {
+					t.Errorf("Expected ToAddresses %v, got %v", desiredRule2.ToAddresses, actualRule.ToAddresses)
+				}
+				if !actualRule.TargetMembers.Equal(desiredRule2.TargetMembers) {
+					t.Errorf("Expected Pods %v, got %v", desiredRule2.TargetMembers, actualRule.TargetMembers)
+				}
+			}
+		// provide enough timeout as the rate limiter will retry rule reconciling after minRetryDelay.
+		case <-time.After(time.Millisecond * 6000):
+			t.Fatal("Expected two rule updates, got timeout")
+		}
+	}
+	np1 := controller.GetNetworkPolicies(&querier.NetworkPolicyQueryFilter{SourceName: policy1.SourceRef.Name, Namespace: policy1.SourceRef.Namespace})
+	np2 := controller.GetNetworkPolicies(&querier.NetworkPolicyQueryFilter{SourceName: policy2.SourceRef.Name, Namespace: policy2.SourceRef.Namespace})
+	assert.ElementsMatch(t, policy1.Rules, np1[0].Rules)
+	assert.ElementsMatch(t, policy1.AppliedToGroups, np1[0].AppliedToGroups)
+	assert.ElementsMatch(t, policy2.Rules, np2[0].Rules)
+	assert.ElementsMatch(t, policy2.AppliedToGroups, np2[0].AppliedToGroups)
+	assert.Equal(t, 2, controller.GetNetworkPolicyNum())
+	assert.Equal(t, 2, controller.GetAddressGroupNum())
+	assert.Equal(t, 1, controller.GetAppliedToGroupNum())
+}
```

Comment:
This new test doesn't seem covering what the patch is fixing (I think it would succeed even without the fix). It only verifies controller will reconcile individual rules upon the BatchInstall failure, which is essentially same as a unit test for `processAllItemsInQueue` that verifies rules are added to the queue upon BatchInstall failure, except it does more check for individual rule check that have been covered by other unit tests. 
And it seems the way we fake reconciler is not extensible as it requires a seperate struct for each scenario in this fashion. I think we should switch to go mocking like other interfaces when the current mockReconciler doesn't work. 
Since this test is not related to the fix and we will need to backport it, I would suggest removing it from this PR and only merge the fix first. We can discuss how to improve the coverage as follow-up.

Reply:
done